controller/api/destination/server.go

```diff
@@ -371,30 +376,114 @@ func (s *server) sendEndpointProfile(stream pb.Destination_GetProfileServer, pod
+		addr.MetricLabels["namespace"] = pod.Namespace
+
+		// Wrap the addr with the proxy port so that when a protocol hint is
+		// set for opaque traffic, it can properly target the proxy port
+		// instead of the application port.
+		proxyPort, err := getInboundPort(&pod.Spec)
+		if err != nil {
+			return err
+		}
+		endpoint = &Endpoint{
+			addr:      addr,
+			proxyPort: proxyPort,
+		}
+
+		portSpec = &policyPb.PortSpec{
+			Workload: fmt.Sprintf("%s:%s", pod.Namespace, pod.Name),
+			Port:     port,
+		}
 	}
 
-	// Send the default profile without subscribing for future updates. The
-	// profile response will also include an endpoint if the IP (or hostname)
-	// sent in the profile request maps to a pod.
+	// Send the default profile. If this is the profile for a pod, subscribe
+	// for updates from the policy server which can influence whether or not
+	// the destination is treated as opaque.
 	translator := newProfileTranslator(stream, log, "", port, endpoint)
 
-	// If there are opaque ports then update the profile translator
-	// with a service profile that has those values
+	// Send the initial destination profile so that the translator does not
+	// get blocked waiting for policy server updates. After this initial
+	// profile is sent, it will be subscribed for updates if the destination
+	// is for a pod.
 	if len(opaquePorts) != 0 {
+		// Send a destination profile update that includes the opaque
+		// ports.
 		sp := sp.ServiceProfile{}
 		sp.Spec.OpaquePorts = opaquePorts
 		translator.Update(&sp)
 	} else {
 		translator.Update(nil)
 	}
 
+	if portSpec != nil {
+		// Create a client which watches for changes to the inbound policy for
+		// portSpec.
+		client, err := s.policyClient.WatchPort(stream.Context(), portSpec)
+		if err != nil {
+			return err
+		}
+
+		// Receive policy server updates for portSpec.
+		//
+		// Running this in the background allows it to handle Recv errors when
+		// the stream context is canceled. When canceled, this routine will
+		// complete.
+		updates := make(chan *policyPb.Server)
+		go func() {
+			for {
+				update, err := client.Recv()
+				if err != nil {
+					log.Debugf("Shutting down policy server updates for %s:%d", portSpec.Workload, portSpec.Port)
+					break
+				}
+				updates <- update
+			}
+		}()
+
+		// Wait for destination server shutdown, stream cancellation, or
+		// policy server updates.
+		//
+		// When an update is available, a new service profile may be created
+		// which results in a new destination profile being sent on the
+		// stream.
+		var overridden bool
+		for {
+			select {
+			case <-s.shutdown:
+				return nil
+			case <-stream.Context().Done():
+				log.Debugf("GetProfile %s:%s:%d cancelled", pod.Namespace, pod.Name, port)
+				return nil
+			case update := <-updates:
+				if update.GetProtocol().GetOpaque() != nil {
+					if _, ok := opaquePorts[port]; !ok {
+						opaquePorts[port] = struct{}{}
+						overridden = true
```

Comment:
I find it very confusing to maintain this state keeping track of whether this has been overridden and also plumbing similar statekeeping through the ProfileTranslator.

What if, instead, we simply update the translator's endpoint?  i.e. when we get a policy update we use that to build a new opaquePorts set and a new WeightedAddr:

```
opaquePortsWithServerProtocol := make(map[uint32]struct{})
for k, v := range opaquePorts {
  opaquePortsWithServerProtocol[k] = v
}
if update.GetProtocol().GetOpaque() != nil {
  opaquePortsWithServerProtocol[port] = struct{}{}
}
newEndpoint, err = toWeightedAddr(podSet.Addresses[podID], opaquePortsWithServerProtocol, skippedInboundPorts, s.enableH2Upgrade, s.identityTrustDomain, s.controllerNS, s.log)
```

then we can update the translator's endpoint in place

```
*endpoint = *newEndpoint
```

and then send an update

```
if len(opaquePortsWithServerProtocol) != 0 {
// Send a destination profile update that includes the opaque
// ports.
		sp := sp.ServiceProfile{}
		sp.Spec.OpaquePorts = opaquePortsWithServerProtocol
		translator.Update(&sp)
	} else {
		translator.Update(nil)
	}
```

Doing it this way should avoid us needing to make any changes to the profile translator code at all.

Reply:
Great suggestion this worked well. I was able to remove all the changes from the profile translator.